src/audio/codec_adapter/codec/waves.c

```diff
@@ -0,0 +1,681 @@
+	if (ret)
+		goto err;
+
+	comp_dbg(dev, "waves_codec_prepare() done");
+	return 0;
+
+err:
+	comp_err(dev, "waves_codec_prepare() failed %d", ret);
+	return ret;
+}
+
+int waves_codec_process(struct comp_dev *dev)
+{
+	int ret;
+	struct codec_data *codec = comp_get_codec(dev);
+	struct waves_codec_data *waves_codec = codec->private;
+
+	comp_dbg(dev, "waves_codec_process() start");
+
+	MaxxStream_t *i_streams[1] = { &waves_codec->i_stream };
+	MaxxStream_t *o_streams[1] = { &waves_codec->o_stream };
+	MaxxStatus_t status;
+	uint32_t num_input_samples = waves_codec->buffer_samples;
+
+
+	// here input buffer should always be filled up as requested
+	// since noone updates it`s size except code in prepare
+	// kinda odd, but this is how codec adapter operates
+	// on the other hand there is available/produced counters in cpd, so we check them anyways
+	if (codec->cpd.avail != waves_codec->buffer_bytes) {
+		comp_warn(dev, "waves_codec_process() input buffer %d is not full %d",
+			codec->cpd.avail, waves_codec->buffer_bytes);
+		num_input_samples = codec->cpd.avail /
+			(waves_codec->sample_size_in_bytes * waves_codec->i_format.numChannels);
+	}
+
+	waves_codec->i_stream.buffersArray = &waves_codec->i_buffer;
+	waves_codec->i_stream.numAvailableSamples = num_input_samples;
+	waves_codec->i_stream.numProcessedSamples = 0;
+	waves_codec->i_stream.maxNumSamples = waves_codec->buffer_samples;
+
+	waves_codec->o_stream.buffersArray = &waves_codec->o_buffer;
+	waves_codec->o_stream.numAvailableSamples = 0;
+	waves_codec->o_stream.numProcessedSamples = 0;
+	waves_codec->o_stream.maxNumSamples = waves_codec->buffer_samples;
+
+	status = MaxxEffect_Process(waves_codec->effect, i_streams, o_streams);
+	if (status) {
+		ret = status;
+		comp_err(dev, "waves_codec_process() MaxxEffect_Process error %d", status);
+		goto err;
+	}
+
+	codec->cpd.produced = waves_codec->o_stream.numAvailableSamples *
+		waves_codec->o_format.numChannels * waves_codec->sample_size_in_bytes;
+
+	comp_dbg(dev, "waves_codec_process() done");
+	return 0;
+
+err:
+	comp_err(dev, "waves_codec_process() failed %d", ret);
+	return ret;
+}
+
+int waves_codec_apply_config(struct comp_dev *dev)
+{
+	int ret;
+
+	comp_dbg(dev, "waves_codec_apply_config() start");
+	ret =  waves_effect_config(dev, CODEC_CFG_RUNTIME);
+	if (ret)
+		goto err;
+
+	comp_dbg(dev, "waves_codec_apply_config() done");
+	return 0;
+
+err:
+	comp_err(dev, "waves_codec_apply_config() failed %d", ret);
+	return ret;
+}
+
+int waves_codec_reset(struct comp_dev *dev)
+{
+	MaxxStatus_t status;
+	int ret = 0;
+	struct codec_data *codec = comp_get_codec(dev);
+	struct waves_codec_data *waves_codec = codec->private;
+
+	comp_dbg(dev, "waves_codec_reset() start");
+
+	status = MaxxEffect_Reset(waves_codec->effect);
+	if (status) {
+		comp_err(dev, "waves_codec_reset() MaxxEffect_Reset error %d", status);
+		ret = status;
+		goto err;
+	}
+
+	comp_dbg(dev, "waves_codec_reset() done");
+	return ret;
+
+err:
```

Comment:
you dont need this goto, just put it in the if statement

Reply:
fixed